### PR TITLE
Range query.

### LINKED ISSUE #19: SSTable in-memory index data structure
Right now, uses BTree but compaction would benefit from an ADT with O(1) insertions. 

The most important operation for the ADT to support is the "bisect" lookup. With a vector, this would be O(logn). What is it right now with the BTree?

### LINKED ISSUE #23: implementing LSM::scan
It would be nice to have a `scan` method on the key-value store. This is an issue to track its implementation.

### PR DESCRIPTION
**Pull request by ysono**

### CODE REVIEW

File: src/storage/db.rs [hunk 1]
```diff
@@ -32,4 +32,22 @@ impl DB {
             _ => Ok(None),
         }
     }
+
+    pub fn get_range(
+        &self,
+        k_lo: &Option<PrimaryKey>,
+        k_hi: &Option<PrimaryKey>,
```
Comment (ysono) [inline src/storage/db.rs:39]: This ought to be `Option<&PrimaryKey>`. It'll be changed in the next PR.

File: src/storage/lsm.rs [hunk 2]
```diff
@@ -160,4 +163,74 @@ where
...
+                .zip(std::iter::repeat(mt_i))
+            })
+            .kmerge_by(|((a_k, _a_v), a_i), ((b_k, _b_v), b_i)| {
+                a_k < b_k || (a_k == b_k && a_i > b_i)
+            })
+            .unique_by(|((k, _v), _mt_i)| k.clone())
+            .map(|(kv, _mt_i)| kv);
+
+        let mut ssts_iter = ssts_iter.peekable();
+        let mut mts_iter = mts_iter.peekable();
```
Comment (ysono) [inline src/storage/lsm.rs:205]: The reason we separate out one iter for all SSTables and one iter for both Memtables is that the type of the iterators are different.

- The iterators on SSTables are all `KeyValueIterator`.
- The iterators on Memtables are both `BTreeMap<K, V>::Iter` (or something like that)

To put these two different types into one vec requires `Vec<Box<dyn Iterator<Item = ...>>>`. This makes it cumbersome.

In addition, the iterators' `Item` types are different:

- The iterators on SSTables have `Item = Result<(K, V)>`
- The iterators on Memtables have `Item = (&K, &V)`

For these reasons, it's simpler to keep them separate.

**Comment by ysono** [comment]
Note, we are using `KeyValueIterator` for all operations that merges across multiple SSTables. This is inefficient, as we are unnecessarily deserializing values. This can be addressed later.

I'm not sure if the native + itertools abstractions of iterators can handle this use pattern. I actually tried using some ugly raw pointer passing to try to give access to the iterator so that you could call `iter.get_value()` on the iterator. This is a bad idea! Seg fault occurs. It shows that rust's ownership transfer does really cause a c++ `std::move()` kind of moving.

File: src/storage/lsm.rs [hunk 3]
```diff
@@ -126,38 +125,112 @@ where
...
+
+        let out_iter_fn = move || -> Option<Result<(K, V)>> {
```
Comment (btc) [inline src/storage/lsm.rs:207]: nice

**Review by btc — APPROVED** [review]

**Comment by btc** [comment]
> Note, we are using KeyValueIterator for all operations that merges across multiple SSTables. This is inefficient, as we are unnecessarily deserializing values. This can be addressed later.

It's worth benchmarking to determine the cost associated with deserialization. With respect to inducing latency, it might be tiny compared to other costs (such as disk accesses). With respect to memory usage, it might be small since there's theoretically only a constant size allocation. With respect to CPU cycles, it might be significant. But worth measuring when contemplating decisions that treat the cost as a material constraint which ought to influence code organization and system architecture.

**Comment by ysono** [comment]
:+1:

**Pull request merged**